File: java-samples/Features/observablestates-tradereporting/contracts/src/main/java/net/corda/samples/observable/contracts/HighlyRegulatedContract.java
package net.corda.samples.observable.contracts;

import net.corda.core.contracts.CommandData;
import net.corda.core.contracts.CommandWithParties;
import net.corda.core.contracts.Contract;
import net.corda.core.identity.AbstractParty;
import net.corda.core.transactions.LedgerTransaction;
import net.corda.samples.observable.states.HighlyRegulatedState;
import org.jetbrains.annotations.NotNull;

import java.util.stream.Collectors;

import static net.corda.core.contracts.ContractsDSL.requireSingleCommand;
import static net.corda.core.contracts.ContractsDSL.requireThat;

public class HighlyRegulatedContract implements Contract {
    public static final String ID = "net.corda.samples.observable.contracts.HighlyRegulatedContract";

    @Override
    public void verify(@NotNull LedgerTransaction tx) throws IllegalArgumentException {
        final CommandWithParties<Commands.Trade> command = requireSingleCommand(tx.getCommands(), Commands.Trade.class);
        requireThat(require -> {

            final HighlyRegulatedState out = tx.outputsOfType(HighlyRegulatedState.class).get(0);
            require.using("The Buyer and the seller cannot be the same entity.",
                    !out.getBuyer().equals(out.getSeller()));
            return null;
        });
    }

    public interface Commands extends CommandData {
        class Trade implements Commands {}
    }
}

File: kotlin-samples/Features/observableStates-tradereporting/contracts/src/main/kotlin/net/corda/samples/observable/contracts/HighlyRegulatedContract.kt
package net.corda.samples.observable.contracts

import net.corda.core.contracts.CommandData
import net.corda.core.contracts.Contract
import net.corda.core.contracts.requireSingleCommand
import net.corda.core.contracts.requireThat
import net.corda.core.transactions.LedgerTransaction
import net.corda.samples.observable.states.HighlyRegulatedState

class HighlyRegulatedContract : Contract {
    companion object {
        const val ID = "net.corda.samples.observable.contracts.HighlyRegulatedContract"
    }

    override fun verify(tx: LedgerTransaction) {
        // Verification logic goes here.
        val cmd = tx.commands.requireSingleCommand<Commands>()
        val output = tx.outputsOfType(HighlyRegulatedState::class.java)[0]
        when(cmd.value){
            is Commands.Trade -> requireThat {
                "The Buyer and the seller cannot be the same entity." using (!output.buyer.equals(output.seller))
            }
        }
    }

    // Used to indicate the transaction's intent.
    interface Commands : CommandData {
        class Trade : Commands
    }
}